### PR TITLE
48 implement test cases for the script GitHub repo_equest_local.py

### PR DESCRIPTION
**Pull request by tnzmnjm**
- Added test cases for `get_base_repo_url` function in the file `tests/test_github_repo_request_local.py`
- Tests will assert the result of:
  -   Typical GitHub repository URLs with and without '.git' suffix.
  -   URLs pointing to issues and other non-repository pages to test correct base URL extraction.
  -   URLs from other code hosting platforms to ensure versatility.
  -   Edge cases including invalid URLs and URLs with query parameters or fragments.
  -   Empty and None input scenarios to validate function robustness.
  - defined the function `filter_incomplete_urls` in the script `src/github_repo_request_local.py` to be able to write test cases efficiently
- adde a helper function inside the `filter_incomplete_urls` function to handle `None` values of the `repourl`
- changed the function `filter_incomplete_urls` to check if the input url is any type other than string (includes `None`)
- wrote a new parametrised test block (`@pytest.mark.parametrize("data, expected_length"`)  for testing different data types, all entries being non-string types, mixed valid URLs and non-string types
- Added parameterised tests for URL validation with special characters testing for URLs containing unencoded special characters such as single quotes, angle brackets, spaces, and double quotes, to verify that they are correctly identified as invalid.
- I am also working on adding CLI arguments for the input and output data

### CODE REVIEW

File: src/github_repo_request_local.py [hunk 1]
```diff
@@ -136,182 +160,219 @@ def get_last_commit_hash(repo_dir: Path):
...
-            except subprocess.CalledProcessError as e:
-                logger.error(
-                    f"Failed to clone the repo: {repo_name}." f" " f"Exception: {e}"
-                )
-                df.at[index, "testfilecountlocal"] = -1
+        # Cleanup based on user's command-line option
+        if (
+            not args.keep_clones
+        ):  # If user did not specify --keep-clones, delete the directory
+            shutil.rmtree(clone_dir)
```
Comment (julianharty) [inline src/github_repo_request_local.py:360]: It's probably worth logging whether the local repos were kept, or deleted, so that someone reading the log knows what happened.

File: tests/test_github_repo_request_local.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    # URLs with query parameters or fragments
+    assert (
+        get_base_repo_url("https://github.com/getdnsapi/stubby.git#readme")
+        == "https://github.com/getdnsapi/stubby"
+    )
+
+    assert (
+        get_base_repo_url("https://github.com/getdnsapi/stubby.git?branch=main")
+        == "https://github.com/getdnsapi/stubby"
```
Comment (julianharty) [inline tests/test_github_repo_request_local.py:61]: This is a test case that provided me some insights, thank you. In this example the branch is `main` which is the one we'd get with a git clone. In future we might want/need to clone another branch. Our current code clones the default (`main`|`master`)  branch only. 

Nothing to do at this stage, more an observation and a thank you for this example.

File: tests/test_github_repo_request_local.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    [
+        # Test with an empty DataFrame
+        ({"repourl": []}, 0, []),
+        # DataFrame without the 'repourl' column
+        ({"other_column": ["data1", "data2"]}, 2, []),
+        # DataFrame with various types of URLs including None and empty strings
+        (
+            {
+                "repourl": [
+                    "https://github.com/user/repo",
```
Comment (julianharty) [inline tests/test_github_repo_request_local.py:76]: As a detail, GitHub uses the term `owner` rather than `user` and I think `owner` is more reflective of/for many of the repos we're processing as they're owned by an organisation rather than the property of an individual person. So let's use `owner` please unless there are good reasons to prefer `user`.

**Review by julianharty — COMMENTED** [review]
Good work. Some very minor tweaks to consider to improve the readability of the code.

File: src/github_repo_request_local.py [hunk 4]
```diff
@@ -136,182 +160,219 @@ def get_last_commit_hash(repo_dir: Path):
...
-        df["testfilecountlocal"] = -1  # Initialise if first run
-    else:
-        logger.error(f"CSV file not found at {csv_file_path}.")
+    # Helper function to determine if a URL is complete
+    def is_complete_url(url):
+        # Check if the url is not a string
+        if not isinstance(url, str):
+            return False
+        parts = url.rstrip("/").split("/")
+        return len(parts) >= 5
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py:194]: This is done.

File: src/github_repo_request_local.py [hunk 5]
```diff
@@ -136,182 +160,219 @@ def get_last_commit_hash(repo_dir: Path):
...
-    df = pd.read_csv(updated_csv_path)
-else:
-    csv_file_path = repo_root / "data" / "original_github_df.csv"
+    Returns:
+        pd.DataFrame: DataFrame with rows containing complete URLs.
+    """
+    if "repourl" not in df.columns:
+        logger.error(
+            "DataFrame does not contain 'repourl' column. Returning the input "
+            "dataframe."
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py:183]: - rising a value error `raise ValueError("DataFrame must contain a 'repourl' column.")`
- updated the `docstring`

File: src/github_repo_request_local.py [hunk 6]
```diff
@@ -136,182 +160,219 @@ def get_last_commit_hash(repo_dir: Path):
...
+    # Identify rows with incomplete URLs using the helper function
+    incomplete_urls = df[~df["repourl"].apply(is_complete_url)]
 
     # Log the incomplete URLs
     if not incomplete_urls.empty:
         logger.info(
-            f"{len(incomplete_urls)} incomplete GitHub URLs found and "
-            f"will be excluded:"
+            f"{len(incomplete_urls)} incomplete GitHub URLs found and will be "
+            f"excluded:"
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py:203]: This is done.

File: src/github_repo_request_local.py [hunk 7]
```diff
@@ -136,182 +160,219 @@ def get_last_commit_hash(repo_dir: Path):
...
 
-    df = df[df["repourl"].apply(lambda x: len(x.rstrip("/").split("/")) >= 5)]
+        if csv_file_path.exists():
+            df = pd.read_csv(csv_file_path)
+            df["testfilecountlocal"] = -1  # Initialise if first run
+        else:
+            logger.error(f"CSV file not found at {csv_file_path}.")
 
+    # Filter out rows where the URL doesn't have a repository name (83 rows)
+    df = filter_incomplete_urls(df)
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/github_repo_request_local.py:244]: - This is done. 
- I have also updated the `tests/test_github_repo_request_local.py` to handle this exception using `pytest.raises`

File: src/github_repo_request_local.py (hunk 1 above)
Comment (tnzmnjm) [reply src/github_repo_request_local.py:360]: Added a `logger.info` before deleting the directory

File: tests/test_github_repo_request_local.py (hunk 2 above)
Comment (tnzmnjm) [reply tests/test_github_repo_request_local.py:61]: Thank you.

**Review by julianharty — APPROVED** [review]
Thank you for the most recent changes. You're welcome to merge them into main.
